### DISCUSSION TITLE
Shader uniforms capacity

### DISCUSSION THREAD

**Discussion by Zyko0** [hajimehoshi/ebiten#1625] [Q&A] [answered]
Hello, I read some opengl about the limitations of passing a huge amount of data as uniforms to a shader.
But in my application I do have specific needs of describing a scene to be rendered on GPU, I read that I do have two options:
* Creating a texture containing my data
Using a texture with ebiten, I don't know of a way to keep the precision of my `float` values.

* Using UBO (https://www.khronos.org/opengl/wiki/Uniform_Buffer_Object) but I don't know if it's too much opengl-related, or if it can be implemented/accessed in Ebiten.

As an example, let's say I have a single shader rendering a scene by computing 800 complex shapes analytically (described as `vec2 position, vec2 size` for example), this would require too many uniforms, if the scene is rendered like:
```go
DrawRectShader(screenWidth, screenHeight, sceneShader, optsUniformsOf800shapes)
```

Currently I'm going with:
```go
for _, e := range entities {
   DrawRectShader(e.Width, e.Height, entityShader, optsWithFewUniforms)
}
```
But this is really gpu intensive, and shouldn't require multiple shader invocations.
It also goes against the automatic batching logic, which can't apply here..

How should I solve this problem with these constraints using Ebiten ?

Thanks in advance ! 😄

**Comment by hajimehoshi** [comment, marked as answer]
Disclaimer: I'm not an expert shading-language programmer.

This depends on how much uniform data you need, but in general, using a texture as a data array is the only way to go so far.

> Creating a texture containing my data
> Using a texture with ebiten, I don't know of a way to keep the precision of my float values.

Yes, the precision might matter. Could you give an example of the shapes data you want to pass to the shader?

**Reply by Zyko0** [reply]
Yes, something like (this is random pseudo code, but it's my idea without `struct`):
```go
// Let's imagine I want to be able to draw 300 shapes at max
// Positions where the shapes need to be drawn
var Positions [300]vec2
// Functions enum describing what kind of complex shape it is [1-N]
// 1 for a simple square
// 2 for a circle
// 3 for a shrimp (idk)
var Functions [300]float
// Angles defines where the orientation of the shape
var Angles [300]float
// Sizes or Params, being read differently depending of the shape requirements (radius, width, height, etc..)
var Sizes [300]vec2
// etc..
```

There would be a standard way of defining sprites in textures, but this is the way I want to approach it, having smooth shapes procedurally generated without vertices, allowing for more dynamic things.

> but in general, using a texture as a data array is the only way to go so far.

Yeah, I'm wondering how to..
Am I supposed to `* 255` my `0-1` logic float values in `r,g,b` components of an `ebiten.Image` before passing it too the shader ? 🤔 

edit: But then what about `0-10` or positions values ?

> Disclaimer: I'm not an expert shading-language programmer.

Also, I'm not at all either, just very interested in it haha

**Reply by hajimehoshi** [reply]
Uniform variables can be a slice of floats. Have you tried?

> There would be a standard way of defining sprites in textures, but this is the way I want to approach it, having smooth shapes procedurally generated without vertices, allowing for more dynamic things.

Actually creating vertices from your data is a usual way. So is passing the data as non vertices required for smooth shape rendering?

> Am I supposed to * 255 my 0-1 logic float values in r,g,b components of an ebiten.Image before passing it too the shader ? 🤔

Yes. 8bit values should be a safe accuracy.

> edit: But then what about 0-10 or positions values ?

I don't understand. 0-10 can be converted to 0-10/255.0.

**Reply by Zyko0** [reply]
> Uniform variables can be a slice of floats. Have you tried?

Yes this is what I'm currently using, but there is a limit on the number of uniforms to pass depending on the GPU

> Actually creating vertices from your data is a usual way. So is passing the data as non vertices required for smooth shape rendering?

Not necessarily I guess, but I'm not sure I can accomplish what I need with meshes in my scenario, and creating elipsoid shapes combined with each other to build more complex shapes (https://iquilezles.org/www/articles/distfunctions2d/distfunctions2d.htm), is the job of the shader and the GPU I believe.

> Yes. 8bit values should be a safe accuracy.

Okay thanks, I'm okay with that !

> I don't understand. 0-10 can be converted to 0-10/255.0.

Yes but I don't know how I would encode them (even a position of 453,512 on screen) if passing them as texture data

edit: I could actually send `pos([2]int) / screenSize([2]int) * 255` and multliply again by a `screenSize vec2` once in shader

**Reply by hajimehoshi** [reply]
> even a position of 453,512 on screen

I'd split a 16bit integer into two 8bit integers and then use 2 floats for 1 value.

> I could actually send pos([2]int) / screenSize([2]int) * 255 and multliply again by a screenSize vec2 once in shader

Nice! I hope this works with any GPUs.

**Reply by Zyko0** [reply]
> I'd split a 16bit integer into two 8bit integers and then use 2 floats for 1 value.

Oh yeah that's a good idea, I believe this would keep enough precision, thank you !

> Nice! I hope this works with any GPUs.

Yes, if everything is encoded into a texture, I guess it should work anywhere !

I'll try going this road, thanks again !

I'm just wondering how much will "cost" all this texture encoding, as well as `imgTextureAt()` calls in shader, if data changes at each `Draw`

**Reply by hajimehoshi** [reply]
Note: If we want to treat 'unnormalized' textures whose elements are [0-255] instead of [0.0-1.0], `GL_EXT_texture_integer` seems available in OpenGL, but I don't think this is available in WebGL.

https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_integer.txt

**Reply by hajimehoshi** [reply]
> I'm just wondering how much will "cost" all this texture encoding, as well as imgTextureAt() calls in shader, if data changes at each Draw

The cost is `ReplacePixels` to update the pixels, and that's all I guess.

**Reply by Zyko0** [reply]
Yeah maybe bandwith and texture lookup with a bunch `imgTextureAt()` to decode the scene at each `Draw/Fragment` pass if there is a lot of data.
But I will give a try ! 😅

**Reply by hajimehoshi** [reply]
> Yeah maybe bandwith and texture lookup with a bunch imgTextureAt() to decode the scene at each Draw/Fragment pass if there is a lot of data.

Ah, that sounds an issue. But however you passed the data, it'd be inevitable to access the data.
